ayon_server/settings/overrides.py

```diff
@@ -169,42 +169,61 @@ def extract_overrides(
-        for key in path:
-            if key not in current:
-                break
-            if key == path[-1]:
-                del current[key]
-                break
-            current = current[key]
 
-    def crawl(obj, ovr, ex, target):
-        for name, _field in obj.__fields__.items():
-            child = getattr(obj, name)
-            if isinstance(child, BaseSettingsModel) and not child._isGroup:
-                if child.dict() != ovr.dict()[name] or (name in ex):
-                    target[name] = {}
-                    crawl(child, getattr(ovr, name), ex.get(name, {}), target[name])
+    result: dict[str, Any] = {}
+    print("Explicit pins", explicit_pins)
+    print("Explicit unpins", explicit_unpins)
+
+    def crawl(
+        original_object: BaseSettingsModel,
+        new_object: BaseSettingsModel,
+        existing_overrides: dict[str, Any],
+        target: dict[str, Any],
+        path: list[str],
+    ):
+        for field_name in original_object.__fields__.keys():
+            old_child = getattr(original_object, field_name)
+            new_child = getattr(new_object, field_name)
+            rpath = [*path, field_name]
+            print("Crawling", rpath)
+
+            if isinstance(old_child, BaseSettingsModel) and not old_child._isGroup:
+                if rpath in (explicit_pins or []):
+                    print("Pinning", rpath)
+                    target[field_name] = new_child.dict()
+
+                elif old_child.dict() != new_child.dict() or (
+                    field_name in existing_overrides
+                ):
+                    target[field_name] = {}
+                    crawl(
+                        original_object=old_child,
+                        new_object=new_child,
+                        existing_overrides=existing_overrides.get(field_name, {}),
+                        target=target[field_name],
+                        path=rpath,
+                    )
             else:
-                if getattr(ovr, name) != getattr(obj, name) or (name in ex):
-                    target[name] = ovr.dict()[name]
+                if (
+                    old_child != new_child
+                    or (field_name in existing_overrides)
+                    or (rpath in (explicit_pins or []))
+                ):
+                    # we need to use the original object to get the default value
+                    # because of the array handling
+                    # old_value = original_object.dict()[field_name]
+                    new_value = new_object.dict()[field_name]
+                    target[field_name] = new_value
 
-    crawl(default, overriden, existing_overrides, result)
+    crawl(default, overriden, existing_overrides, result, [])
+
+    # remove paths that are explicitly unpinned
+
+    for path in explicit_unpins or []:
+        current = result
+        for key in path[:-1]:
+            current = current[key]
+        current.pop(path[-1], None)
```

Comment:
This may be easier to read with a dedicated function:
```python
def pop_path(data: dict, path: List[str]) -> Any:
    """Pop a child path nested deep in a dict (if it exists).

    If the key or any of its parents do not exist this will
    return None.

   Arguments:
       data (dict): The data to remove from.
       path (List[str]): The nested keys to pop.

    Returns:
        Any: The popped value, if any.
"""
    current = data
    parents, pop_key = keys.pop()
    for key in parents:
        current = data.get(key, {})
        if not current:
            return
        if not isinstance(current, dict):
            raise TypeError(f"Parent value for key {key} is not a dict: {current}")
    return current.pop(pop_key, None)
```

So that here the logic is just:
```python
for path in explicit_unpins:
    pop_path(result, path)
```

Reply:
i realized, there already is a function to do this in ayon_server.utils :D

Reply:
Even better!